contracts/exchangeIssuance/IssueRedeem.sol

```diff
@@ -0,0 +1,192 @@
+    IUniswapV2Router02 private sushiRouter;
+    address private sushiFactory;
+
+    IBasicIssuanceModule private basicIssuanceModule;
+    address constant private WETH = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+    uint256 constant private MAX_INT = 2**256 - 1;
+
+    /* ============ Constructor ============ */
+
+    constructor() public {
+        uniFactory = address(0x5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f);
+        uniRouter = IUniswapV2Router02(address(0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D));
+
+        sushiFactory = address(0xC0AEe478e3658e2610c5F7A4A2E1777cE9e4f2Ac);
+        sushiRouter = IUniswapV2Router02(address(0xd9e1cE17f2641f24aE83637ab66a2cca9C378B9F));
+
+        basicIssuanceModule = IBasicIssuanceModule(address(0xd8EF3cACe8b4907117a45B0b125c68560532F94D));
+        IERC20(WETH).approve(address(uniRouter), MAX_INT);
+        IERC20(WETH).approve(address(sushiRouter), MAX_INT);
+    }
+
+    /* ============ External Functions ============ */
+
+    receive() external payable {}
+
+    /**
+     * Runs all the necessary approval functions required before issuing
+     * or redeeming an index. This function must only be called before the first time
+     * this smart contract is used on any particular index, or when a new token is added
+     * to the index.
+     *
+     * @param _index    Address of the index being initialized
+     */
+    function initApprovals(address _index) external {
+        ISetToken.Position[] memory positions = ISetToken(_index).getPositions();
+        for (uint256 i=0; i<positions.length; i++) {
+            IERC20 token = IERC20(positions[i].component);
+            token.approve(address(uniRouter), MAX_INT);
+            token.approve(address(sushiRouter), MAX_INT);
+            token.approve(address(basicIssuanceModule), MAX_INT);
+        }
+    }
+
+    /**
+     * Redeems an Index and sells the underlying tokens for ETH using Uniswap
+     * or Sushiswap.
+     *
+     * @param _index    Address of the index being redeemed
+     * @param _amount   The amount of the index to redeem
+     */
+    function redeem(address _index, uint256 _amount) external {
+        ISetToken setToken = ISetToken(_index);
+        setToken.transferFrom(msg.sender, address(this), _amount);
+        basicIssuanceModule.redeem(ISetToken(_index), _amount, address(this));
+        liquidateSetPositions(_index);
+    }
+
+    /**
+     * Issues an Index by using swapping for the underlying tokens using ETH on Uniswap
+     * or Sushiswap. msg.value must be equal to the maximum price in ETH that you are
+     * willing to pay. Excess ETH is refunded.
+     *
+     * @param _index    Address of the index being issued
+     * @param _amount   Amount of the index to issue
+     */
+    function issue(address _index, uint256 _amount) external payable {
+        ISetToken.Position[] memory positions = ISetToken(_index).getPositions();
+        uint256[] memory tokenAmounts = new uint256[](positions.length);
```

Comment:
For better encapsulation, this logic and the logic in the for-loop can be put into an internal/private function